src/java/org/apache/cassandra/service/accord/interop/AccordInteropExecution.java

```diff
@@ -234,123 +227,125 @@ public EndpointsForToken forNonLocalStrategyTokenRead(ClusterMetadata doNotUse,
         node.send(id, commit, executor, new AccordInteropRead.ReadCallback(id, to, message, callback, this));
     }
 
     @Override
     public void sendReadRepairMutation(Message<Mutation> message, InetAddressAndPort to, RequestCallback<Object> callback)
     {
-        checkArgument(message.payload.allowsPotentialTransactionConflicts());
+        checkArgument(message.payload.potentialTxnConflicts().allowed);
+        checkArgument(message.payload.getTableIds().size() == 1);
         Node.Id id = endpointMapper.mappedId(to);
+        Participants<?> readScope = Participants.singleton(txn.read().keys().domain(), new TokenKey(message.payload.getTableIds().iterator().next(), message.payload.key().getToken()));
         AccordInteropReadRepair readRepair = new AccordInteropReadRepair(id, executes, txnId, readScope, executeAt.epoch(), message.payload);
         node.send(id, readRepair, executor, new AccordInteropReadRepair.ReadRepairCallback(id, to, message, callback, this));
     }
 
-    private List<AsyncChain<Data>> keyReadChains(int nowInSeconds, Dispatcher.RequestTime requestTime)
+    private List<AsyncChain<Data>> readChains(Dispatcher.RequestTime requestTime)
     {
-        TxnKeyRead read = (TxnKeyRead) txn.read();
-        List<AsyncChain<Data>> results = new ArrayList<>();
+        TxnRead read = (TxnRead) txn.read();
         Seekables<?, ?> keys = txn.read().keys();
+        switch (keys.domain())
+        {
+            case Key:
+                return keyReadChains(read, keys, requestTime);
+            case Range:
+                return rangeReadChains(read, keys, requestTime);
+            default:
+                throw new IllegalStateException("Unhandled domain " + keys.domain());
+        }
+    }
+
+    private List<AsyncChain<Data>> keyReadChains(TxnRead read, Seekables<?, ?> keys, Dispatcher.RequestTime requestTime)
+    {
+        ClusterMetadata cm = ClusterMetadata.current();
+        List<AsyncChain<Data>> results = new ArrayList<>();
+        keys.forEach(key -> {
+                         read.forEachWithKey((PartitionKey) key, fragment -> {
+                             SinglePartitionReadCommand command = (SinglePartitionReadCommand) fragment.command();
+
+                             // This should only rarely occur when coordinators start a transaction in a migrating range
+                             // because they haven't yet updated their cluster metadata.
+                             // It would be harmless to do the read, because it will be rejected in `TxnQuery` anyways,
+                             // but it's faster to skip the read
+                             AccordClientRequestMetrics metrics = txn.kind().isWrite() ? accordWriteMetrics : accordReadMetrics;
+                             // TODO (required): This doesn't use the metadata from the correct epoch
```

Comment:
this pattern is all over the place, so i do wonder about the intended semantic... if a range is part of accord at epoch=N but not epoch=N+1 and we are executing in epoch=N but locally we know epoch=N+1 should we allow progress?  If so we need similar logic as drop tables that makes the TCM change multi staged and blocks waiting on everything allowed for that epoch.

Reply:
Live migration is built around the decision making for whether to execute on Accord or not being correctly tied to TCM epochs and the `TableMigrationState` stored in TCM containing what ranges are migrating. It's not like add/drop table because there is no secondary set of data that it is dependent on like `Schema.instance` it's completely self contained in that in the cluster metadata.

Accord picks an execution timestamp and that determine the epoch of a transaction. Paxos has each node vote on it as part of running Paxos. Non-Paxos/Non-Accord writes are both voting on whether the write needs to be retried on Accord while also blocking local application of the write at nodes where Accord may have already started reading.